File: hash-table/RansomNote.cpp
#include <unordered_map>

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        std::unordered_map<char, int> counter;

        for (auto c : magazine) {
            counter[c]++;
        }

        for (char c : ransomNote) {
            if (counter.find(c) == counter.end() || counter[c] == 0)
                return false;

            counter[c]--;
        }

        return true;
    }
};

File: hash-table/ransom_note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        counter = {}

        for c in magazine:
            counter[c] = counter.get(c, 0) + 1

        for c in ransomNote:
            if c not in counter or counter[c] == 0:
                return False
            counter[c] -= 1

        return True
